### ISSUE TITLE
Support later versions of Node (or make version configurable)

### ISSUE THREAD

**Issue by blittle** [stackblitz/webcontainer-core#560]
**Is your feature request related to a problem? Please describe:**
At Shopify we heavily use https://hydrogen.new for sharing and promoting the [Hydrogen Framework](https://github.com/Shopify/hydrogen). We recently made changes that require Node v16.5.0 or greater. If there's no way to configure stackblitz to use Node 16, we'll probably revert those changes until this issue is resolved.

**Describe the solution you'd like:**
We'd like a way to configure which version of NodeJS the stackblitz container runs with.

**Describe alternatives you've considered:**
Our alternative is to 1) not use stackblitz or 2) revert our Node 16 specific features, though this will have to be tempoarary because the features are required for our 1.0 launch.

**Additional context:**
Related issue: https://github.com/stackblitz/core/issues/1622

**Comment by EricSimons** [comment]
Hey @blittle- thanks for reporting this. Node 16 support is on our roadmap for Q1 and WIP on our side. Supporting multiple Node versions is a big change in webcontainer so this will likely take a few months to land, but after that it will be much easier for us to add support for subsequent versions of Node (and other langs/runtimes).

I'm curious what specific features of Node 16 that Hydrogen is relying on that isn't present in 14? Depending on what those are there might be something we can do from our side as a temp solution, or perhaps some sort of polyfill approach in Hydrogen core that would allow Node 14 users to also run Hydrogen v1?

**Comment by blittle** [comment]
Hi @EricSimons thank you for the response! It's great to hear that Node 16 is WIP. The specific features we rely upon are the [`ReadableStream` API](https://nodejs.org/api/webstreams.html#example-readablestream). Though I'm currently exploring alternatives.

**Comment by trentprynn** [comment]
Just chiming in to add that I'm excited for this feature and add a reason it will be super useful to me!

Currently I primarily use stackblitz as a way to share bug reproductions (it's awesome for this, thank you!) and I'm having a problem I'm relatively sure is related node@16 that I cannot reproduce / share this way because stackblitz only currently only supports node@14

**Comment by rayfoss** [comment]
For me it's webcrypto... completely missing in Node 14
```
import Koa from 'koa'
import logger from 'koa-logger'
import Router from '@koa/router'
import { webcrypto } from 'crypto'

const crypto = webcrypto as unknown as Crypto

const random = crypto.getRandomValues(new Uint8Array(24))
```


```
import { webcrypto } from 'crypto';
         ^^^^^^^^^
SyntaxError: The requested module 'crypto' does not provide an export named 'webcrypto'
```

https://stackblitz.com/edit/koa-starter-wychx9?file=src%2Findex.ts

**Comment by jycouet** [comment]
I just found out that we are on node `v14.19.0`, any idea when we will be able to tune this with a parameter?
Thank you

**Comment by rayfoss** [comment]
> I just found out that we are on node `v14.19.0`, any idea when we will be able to tune this with a parameter? Thank you

I'm against taking the burden of supporting multiple versions of node. If someone needs to archive everything, containers are better for that. It would be nice to have the option of using node 17, but what we'll always desperately need is the current LTS, which right now is 16.

Very few projects will break going from 14 to 16. When they do, the fix is usually trivial. Right now there is no production pipeline that will break if we hard switch everyone to Node 16. So lets switch everyone to LTS and forget about 14, it's the easiest, cheapest and best route long term and short term.

**Comment by codepunkt** [comment]
The live servers don't have the latest LTS.

Stackblitz and other platforms are amazing for providing error reproductions - but whenever these error reproductions are connected to specific Node.js versions, you're out of look. It would be a big boost to our usage of Stackblitz if we could choose the Node version we want, including minors and pre-releases.

**Comment by vlucas** [comment]
Re-visiting this issue because **Node 18 has landed** and [includes support for some major web primitives that are now built-in](https://nodejs.org/ko/blog/announcements/v18-release-announce/#new-globally-available-browser-compatible-apis) to Node itself as globals. Things like `fetch`, `FormData`, `Request`, etc. that we use a lot in Hydrogen.

Node 18 also includes [Web Streams](https://nodejs.org/ko/blog/announcements/v18-release-announce/#web-streams-api-experimental) (`ReadableStream`, WriteableStream`, etc.) now built-in the global Node namespace that Hydrogen makes heavy use of for streaming server responses with React 18.

Upgrading to Node 18 and requiring it as the minimum version to use Hydrogen means we would no longer have to use feature detection or polyfill anything. This would improve performance and reduce overall bundle and install size for all Hydrogen apps (as well as many other Node apps too that no longer have to use things like `node-fetch`!).

Do you have any ETA on when we can specify a specific Node version in our own container, or when you can upgrade to Node 18 (the current LTS version)?

**Comment by d3lm** [comment]
Hey there 👋 We are planning to upgrade to Node 18 in the next quarter or two before 16 goes out of maintenance. We keep you posted here.

**Comment by thescientist13** [comment]
@d3lm 
Great to hear!  A lot of great Web APIs landed in Node 18 like native `fetch`, `URL`, `Request` and `Response`  so it would be great to see support for those in Web Containers.  💯

**Comment by d3lm** [comment]
Yea, once upgraded to 18 those should work 👍

**Comment by bluwy** [comment]
Leaving a note that Astro 3.0 (releasing before node16 EOL) will only support node18 and above. So it might temporarily not work in stackblitz as we use some node18 APIs, like `File`, `Headers`, `Request`, `fetch`.

**Comment by nickwinger** [comment]
With the Qwik Framework and server side functions to work, we would also need node 18

**Comment by brophdawg11** [comment]
[Remix v2](https://github.com/remix-run/remix/releases/tag/remix%402.0.0) also upgraded to a minimum of Node 18, so we'd love to see this become configurable (or raised to current LTS)

**Comment by austinm911** [comment]
[T3-Turbo](https://github.com/t3-oss/create-t3-turbo/blob/main/package.json) requires Node.20.10 so unable to get it working on Stackblitz

**Comment by NullVoxPopuli** [comment]
> We are planning to release an upgrade to Node.js 18 in the coming 1 to 2 weeks

Didn't look like this happened?

I opened an issue here: https://github.com/stackblitz/core/issues/2926

webpack is totally broken (if using the thread-loader, due to the libUV issue https://github.com/nodejs/node/issues/49911)

**Comment by SamVerschueren** [comment]
> Didn't look like this happened?

We did do the upgrade from 16 to 18. This answer was not about upgrading to a different 18 minor.

> webpack is totally broken (if using the thread-loader, due to the libUV issue https://github.com/nodejs/node/issues/49911)

I already thought this issue looked familiar :). Apparently you ran into (and reported) it a couple months ago https://github.com/stackblitz/webcontainer-core/issues/1256. I also answered why `thread-loader` is broken in WebContainer.

**Comment by SamVerschueren** [comment]
> Does that also make it easier for devs to get up to speed faster with new LTS releases on release days?

The plan (on paper) is to be able to support new LTS releases as soon as possible. Release day is probably too fast, but let's say within 4-6 weeks.

> Also, do we have an idea of when Node.js 20 (LTS) will be supported?

We're aiming for Q2. We have already done some of the work but had to shift priorities. We will pick it up again very soon.

**Comment by newstart0514** [comment]
I really need it because in recent projects, some NPM packages require an environment with a minimum of NodeJS 20

> > Does that also make it easier for devs to get up to speed faster with new LTS releases on release days?
> 
> The plan (on paper) is to be able to support new LTS releases as soon as possible. Release day is probably too fast, but let's say within 4-6 weeks.
> 
> > Also, do we have an idea of when Node.js 20 (LTS) will be supported?
> 
> We're aiming for Q2. We have already done some of the work but had to shift priorities. We will pick it up again very soon.

**Comment by GeeekyBoy** [comment]
> > We are planning to release an upgrade to Node.js 18 in the coming 1 to 2 weeks
> 
> Didn't look like this happened?
> 
> I opened an issue here: [stackblitz/core#2926](https://github.com/stackblitz/core/issues/2926)
> 
> webpack is totally broken (if using the thread-loader, due to the libUV issue [nodejs/node#49911](https://github.com/nodejs/node/issues/49911))

Good news 🎉
StackBlitz now uses Node.js `v18.20.3` instead of Node.js `v18.18.0`.

**Comment by SamVerschueren** [comment]
Due to priorities, Node.js 20 is planned to be worked on starting end of July.

**Comment by ruckc** [comment]
@SamVerschueren ... and when Node 22 hits LTS in October?

**Comment by SamVerschueren** [comment]
The plan is to (almost) immediately start working on Node.js 22 after Node.js 20 has landed.

**Comment by eslachance** [comment]
Hey guys just reminding you that you're still on Node 18, and still haven't found a way to let us determine the version of node we want. This is honestly a little disappointing.

**Comment by didier** [comment]
Yep, stackblitz is on node 18.

**Comment by dirkluijk** [comment]
Radical idea: what if you can switch runtimes on StackBlitz? E.g. switch from Node to Deno, and also accept JSR as custom NPM registry. I think that would really boost collaboration in open-source JavaScript based projects. 

Let me know if you need any contribution, I’d love to spend some time on this as a public feature.

**Comment by BananaAcid** [comment]
2 Weaks ago, a few of my projects started breaking ... more started following. Due to dependency updates and node being to old. `npm ERR! t._handle.getAsyncId is not a function` ... very unfortunate.

**Comment by jkoefer** [comment]
Any updates on this from the Stackblitz team?

**Comment by ZipBrandon** [comment]
Hi, StackBlitz and @SamVerschueren. Is Node 20 still a priority? Bolt.new is being limited with some frameworks because of the minimum versions of Node needed. Thank you!

**Comment by cweekly** [comment]
Please prioritize support for Node LTS version selection -- and please also be sure to bring this capability to Stackblitz Enterprise ASAP. I've been championing Stackblitz EE at my workplace, and it'd be a real shame if we finally make it happen only to get hung up on being stuck w/ outdated Node. Hopefully the influx of cash from Bolt ($4m ARR in its first month??) will help Stackblitz fund this kind of request -- which IMHO is fundamental. Thanks!

**Comment by rubymignot** [comment]
Impossible to run latest versions of Nuxt in Bolt.diy, do we have an ETA on this?

**Comment by Ashar2shahid** [comment]
not having latest node version is a real bottleneck

**Comment by mrlubos** [comment]
Hey! In my case, even Node 18 doesn't work

```sh
❯ npm install && npm run dev
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@nuxt/kit@3.15.1',
npm WARN EBADENGINE   required: { node: '>=18.20.5' },
npm WARN EBADENGINE   current: { node: 'v18.20.3', npm: '10.2.3' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@nuxt/telemetry@2.6.4',
npm WARN EBADENGINE   required: { node: '>=18.20.5' },
npm WARN EBADENGINE   current: { node: 'v18.20.3', npm: '10.2.3' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@nuxt/vite-builder@3.15.1',
npm WARN EBADENGINE   required: { node: '^18.20.5 || ^20.9.0 || >=22.0.0' },
npm WARN EBADENGINE   current: { node: 'v18.20.3', npm: '10.2.3' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'nuxt@3.15.1',
npm WARN EBADENGINE   required: { node: '^18.20.5 || ^20.9.0 || >=22.0.0' },
npm WARN EBADENGINE   current: { node: 'v18.20.3', npm: '10.2.3' }
npm WARN EBADENGINE }
```

**Comment by NullVoxPopuli** [comment]
Is node 18 bundled in the browsers, and not something stackblitz has control over?

**Comment by vineetdigit** [comment]
New react-router v7 requires Node 20. I can't use stackblitz to create repros.

**Comment by jack3898** [comment]
I have been subscribed to this thread and this request keeps coming back in waves. Something needs to be done :/

**Comment by cweekly** [comment]
Node 18 will leave maintenance in April:  
[https://github.com/nodejs/release#release-schedule](https://github.com/nodejs/release#release-schedule)  

This ticket's history has me concerned about Stackblitz's viability for serious projects. :(

**Comment by EricSimons** [comment]
Hi all! We've been inundated with demand on our new Bolt.new product recently which caused a delay on this, but work resumed the past weeks and we expect to have Node 20 online well before April. Once that's landed we're immediately beginning work on Node 22, and from here on out expect to have a more seamless cadence of having multiple versions available in line with LTS.

PS- we just [raised more funding](https://x.com/boltdotnew/status/1882106655258894390) and are expanding our WebContainer team, so [please apply](https://jobs.lever.co/stackblitz/7ccd2472-1416-4448-8642-e58e9b32e129) if you'd like to work on this with us!! 🙌

**Comment by iamchathu** [comment]
Why doesn't support 22 LTS version yet?

**Comment by acusti** [comment]
@iamchathu see the [latest **update](https://github.com/stackblitz/webcontainer-core/issues/560#issuecomment-2611279611)** right above your question:

> We've been inundated with demand on our new Bolt.new product recently which caused a delay on this, but work resumed the past weeks and we expect to have Node 20 online well before April. Once that's landed we're immediately beginning work on Node 22

**Comment by CollinHerber** [comment]
Has Stackblitz stopped caring about the webcontainer?
We are a few node versions behind at this point and getting repro from consumers has become a problem and looking for other solutions soon if this isn't resolved.

**Comment by SamVerschueren** [comment]
Hello everyone 👋 .

I have an update to share regarding Node 20. Although it's not YET the default when you start a new project, it is already possible to use it 🎉 .

We are looking for feedback and people who test this out so when we make it the default (hopefully very soon), most users won't even notice a difference.

You can either do 1 of these 3 option (whatever flavour you like more).

### engines.node

Adding a `engines.node` field to your `package.json`

```package.json
{
  "name": "my-project",
  "version": "1.0.0",
  "engines": {
    "node": "20.19.0"
  }
}
```

### .nvmrc

If you're more a fan of `.nvmrc`, we got you covered

```
v20.19.0
```

### .tool-versions

Or if you're a `asdf` user, you can also use `.tool-versions`

```
nodejs 20.19.0
```

**Comment by mrlubos** [comment]
Awesome @SamVerschueren! What happens if the defined Node version is 22?

**Comment by SamVerschueren** [comment]
If you define 22 you currently get served with the default version (which is still 18 and will soon become 20). I thought about serving v20 in that case but didn't do it in the end.

**Comment by bmatto** [comment]
Coming across this as I'm trying to use a node package which requires v20 - It's not clear to me _where_ or _which_ package.json should specify the node version. Is it the package.json that is mounted as part of the files? Thank you in advance for any guidance.

**Comment by joaopcm** [comment]
It's still not clear how to use Node.js versions 20 and above in a web container. Could someone share more details about it? Thanks!

**Comment by SamVerschueren** [comment]
If you are using the StackBlitz editor, you just have to add it to the `package.json` in the root of your project.

If you are using [WebContainer API](https://webcontainers.io/), you have to wait a little bit longer until we make it the default. We don't provide an option (yet) to specify which version you want.

**Comment by tadjik1** [comment]
Hi team! What would it take to support current lts (version 22), or future one - 24?
I'm very new to the project and don't know the internals, how exactly the new version is getting into the webcontainers, is it even possible to streamline this process?

**Comment by thescientist13** [comment]
Would be great to get an update on Node 22 LTS support, in particular now that native TypeScript support has come as of 22.18.0

**Comment by ntucker** [comment]
@SamVerschueren is stackblitz.com deprecated? How does one make stackblitz run versions after 18?

**Comment by eslachance** [comment]
> How does one make stackblitz run versions after 18?

I just created a new project with React (Typescript) and I got node v20.19.1 , so clearly higher than 18 is possible... Really wish we actually knew how to switch it because to the contrary of what @SamVerschueren said, neither updating the package.json nor adding a .nvmrc works to switch the version. It's pretty disappointing to see.

**Comment by SamVerschueren** [comment]
By default it runs Node.js 20. You can downgrade to 18 by updating your package.json. We are actively working on upgrading to 22. That will then become the default and we will allow to downgrade to 20 and 18 the same way as you can downgrade to 18 today.

**Comment by SamVerschueren** [comment]
I'll check tomorrow if the downgrading still works or not. If not, I'll fix it.

**Comment by SamVerschueren** [comment]
So I looked into it and for me it still works. One thing to note though is that if you change the version, you will have to refresh the page. It does not change the version on the fly. I've tried the `engines.node` field in `package.json`, the `.nvmrc`, and the `.tool-versions` and they all worked.

So if this doesn't work for you, let me know and maybe even share a small reproduction (make sure the project is public) so I can take a look at what's wrong here.

**Comment by thescientist13** [comment]
Hope to see more investment in Web Containers + Node from Stackblitz on improving upkeep for NodeJS LTS support going forward.  Feels like since Bolt came out Web Containers have taken a backseat and it has been a game of catch-up for a while.  I had to remove Stackblitz demos from my web framework docs pages since Node 22 hasn't been supported for so long.  (I get that it takes work, but as a Stackblitz supporter, it's sad to see it fall behind since I love the technology very much ❤️)

**Comment by SamVerschueren** [comment]
Hello everyone 👋 ! I have some news on the topic of the Node.js 22 release that I've been working on for a while. As of today, you can test Node.js 22 although it's not the default yet! I still need to look into a couple more issues in Firefox before we can make it the default. We would also want to see if there are any bugs like projects that don't work anymore that did work in Node.js 20 and so on.

You can test Node.js 22 by using either one of the following methods.

### engines.node

Adding a `engines.node` field to your `package.json` in the root of your project

```package.json
{
  "name": "my-project",
  "version": "1.0.0",
  "engines": {
    "node": "22.22.0"
  }
}
```

### .nvmrc

If you're more a fan of `.nvmrc`, we got you covered. Just add this file to the root of the project.

```
v22.22.0
```

### .tool-versions

Or if you're a `asdf` user, you can also use `.tool-versions`. Just add this file to the root of the project.

```
nodejs 22.22.0
```

**Comment by mrlubos** [comment]
Great news @SamVerschueren! Are there any updates on how you think about Node 24 and/or support moving forward?

**Comment by SamVerschueren** [comment]
@mrlubos I can't give you any estimation or update on this. The only thing that I can tell is that we usually aim to upgrade when a version goes into maintenance mode. Part of the reason is that at that point the LTS version gets way less updates and thus allows us to support somewhat the latest version within that major version range.

**Comment by ivanm696** [comment]
![Image](https://github.com/user-attachments/assets/0deb5e3b-9796-407d-9acc-50f5f8568cb2)

**Comment by ivanm696** [comment]
![Image](https://github.com/user-attachments/assets/3eb4b348-3a0c-4ede-9263-2f365939aaf7)

**Comment by SamVerschueren** [comment]
@ivanm696 If you post something that doesn't work, please attach reproduction links and information. A screenshot doesn't say anything. It can be very helpful, but at least do some effort if you want other people to spend time on it.

**Comment by dan-lee** [comment]
@SamVerschueren I noticed this works on StackBlitz but not in `@webcontainer/api` directly. Is that expected? Is there a plan to release this soon?

**Comment by SamVerschueren** [comment]
@dan-lee Yes we're planning to make it the default before the end of this week.

**Comment by SamVerschueren** [comment]
Node 22 is now the default. It took a bit longer than expected but we did a gradual rollout over a certain period of time.

**Comment by nigrosimone** [comment]
Angular 22.0.x require NodeJS ^22.22.3
https://angular.dev/reference/versions#actively-supported-versions

current webcontainer is webcontainer is v22.22.0

see https://github.com/stackblitz/webcontainer-core/issues/2106

**Comment by SamVerschueren** [comment]
@nigrosimone This got fixed. We upgraded to `22.22.3` I believe 3 weeks ago.

**Comment by eslachance** [comment]
> [@nigrosimone](https://github.com/nigrosimone) This got fixed. We upgraded to `22.22.3` I believe 3 weeks ago.

Oh hey just in time to be late for Node 24.18 which is currently LTS! 

> Node.js 24 officially transitioned to Long-Term Support (LTS) on October 28, 2025.
